pkg/virt-handler/mount-manager/mount.go

```diff
@@ -0,0 +1,112 @@
+)
+
+//go:generate mockgen -source $GOFILE -package=$GOPACKAGE -destination=generated_mock_$GOFILE
+
+// MountInfo wraps all the mount information
+type MountInfo struct {
+	containerDisksInfo map[string]*containerdisk.DiskInfo
+}
+
+func (info *MountInfo) GetContainerDisksInfo() map[string]*containerdisk.DiskInfo {
+	return info.containerDisksInfo
+}
+
+// MountManager handles all the mount operations required by KubeVirt
+type MountManager interface {
+	Mount(vmi *v1.VirtualMachineInstance) (MountInfo, error)
+	Unmount(vmi *v1.VirtualMachineInstance) error
+	MountsReady(vmi *v1.VirtualMachineInstance, notInitializedSince time.Time) (bool, error)
+	SyncMounts(vmi *v1.VirtualMachineInstance) error
+	SyncUnmounts(vmi *v1.VirtualMachineInstance) error
+	IsMounted(vmi *v1.VirtualMachineInstance, volume string, sourceUID types.UID) (bool, error)
+}
+
+type mountManager struct {
+	containerDiskMounter container_disk.Mounter
+	hotplugVolumeMounter hotplug_volume.VolumeMounter
+}
+
+func NewMounter(virtPrivateDir string, podIsolationDetector isolation.PodIsolationDetector, clusterConfig *virtconfig.ClusterConfig, kubeletPodsDir string) MountManager {
+	return &mountManager{
+		containerDiskMounter: container_disk.NewMounter(podIsolationDetector, clusterConfig, mountutils.NewMountRecorder(filepath.Join(virtPrivateDir, "container-disk-mount-state"))),
+		hotplugVolumeMounter: hotplug_volume.NewVolumeMounter(mountutils.NewMountRecorder(filepath.Join(virtPrivateDir, "hotplug-volume-mount-state")), kubeletPodsDir),
+	}
+}
+
+// MountsReady returns if the mount points are ready to be used
+func (m *mountManager) MountsReady(vmi *v1.VirtualMachineInstance, notInitializedSince time.Time) (bool, error) {
+	// Check container diks are ready to be mounted
+	return m.containerDiskMounter.ContainerDisksReady(vmi, notInitializedSince)
+}
+
+// Mount mounts the volumes managed directly by KubeVirt
+func (m *mountManager) Mount(vmi *v1.VirtualMachineInstance) (MountInfo, error) {
+	disksInfo, err := m.containerDiskMounter.MountAndVerify(vmi)
+	if err != nil {
+		return MountInfo{}, fmt.Errorf("failed to mount container disks: %v", err)
+	}
+
+	attachmentPodUID := types.UID("")
+	if vmi.Status.MigrationState != nil {
+		attachmentPodUID = vmi.Status.MigrationState.TargetAttachmentPodUID
+	}
+	if attachmentPodUID != types.UID("") {
+		if err = m.hotplugVolumeMounter.MountFromPod(vmi, attachmentPodUID); err != nil {
+			return MountInfo{}, fmt.Errorf("failed to mount hotplug volumes: %v", err)
+		}
+
+	} else {
+		if err = m.hotplugVolumeMounter.Mount(vmi); err != nil {
+			return MountInfo{}, fmt.Errorf("failed to mount hotplug volumes: %v", err)
+		}
+	}
+	return MountInfo{
+		containerDisksInfo: disksInfo,
+	}, nil
+}
+
+// Unmount unmounts the volumes managed directly by KubeVirt
+func (m *mountManager) Unmount(vmi *v1.VirtualMachineInstance) error {
+	errCd := m.containerDiskMounter.Unmount(vmi)
+	errHp := m.hotplugVolumeMounter.UnmountAll(vmi)
```

Comment:
Maybe we should return an error right away if unmount for container disks fails? That was the previous behavior.

```go
if err := m.containerDiskMounter.Unmount(vmi); err != nil {
    return err
}
if err := m.hotplugVolumeMounter.UnmountAll(vmi); err != nil {
    return err
}
```

Reply:
Yes, however, I think it will be better to clean-up as much as possible. If there is an error in the container disks clean-up, then we don't perform the clean-up of hotplugged disks, and this could be successful. WDYT?

Reply:
Hm, maybe you're right and for the cleanup it makes sense. Cannot think of any faulty use-case now. I guess it should be safe to call `hotplugVolumeMounter.UnmountAll` multiple times in case smth goes wrong with the container disks.